examples/ex23.cpp

```diff
@@ -0,0 +1,988 @@
+      offsets[2] = fespace->GetTrueVSize();
+      offsets.PartialSum();
+
+      GSSmoother gs00(*PCOpAh.As<SparseMatrix>());
+      BlockDiagonalPreconditioner BlockGS(offsets);
+
+      ScaledOperator gs11(&gs00,(conv == ComplexOperator::HERMITIAN) ? -1.0:1.0);
+
+      BlockGS.SetDiagonalBlock(0,&gs00);
+      BlockGS.SetDiagonalBlock(1,&gs11);
+
+      GMRESSolver gmres;
+      gmres.SetPrintLevel(1);
+      gmres.SetKDim(200);
+      gmres.SetMaxIter(2000);
+      gmres.SetRelTol(1e-5);
+      gmres.SetAbsTol(0.0);
+      gmres.SetOperator(*A);
+      gmres.SetPreconditioner(BlockGS);
+      gmres.Mult(B, X);
+   }
+#endif
+
+   // 15. Recover the solution as a finite element grid function and compute the
+   //     errors if the exact solution is known.
+   a.RecoverFEMSolution(X, b, x);
+
+   if (exact_known)
+   {
+      ComplexGridFunction x_gf(fespace);
+      VectorFunctionCoefficient E_ex_Re(dim, E_exact_Re);
+      VectorFunctionCoefficient E_ex_Im(dim, E_exact_Im);
+      x_gf.ProjectCoefficient(E_ex_Re, E_ex_Im);
+      int order_quad = max(2, 2 * order + 1);
+      const IntegrationRule *irs[Geometry::NumGeom];
+      for (int i = 0; i < Geometry::NumGeom; ++i)
+      {
+         irs[i] = &(IntRules.Get(i, order_quad));
+      }
+
+      double L2Error_Re = x.real().ComputeL2Error(E_ex_Re, irs,
+                                                  pml->GetMarkedPMLElements());
+      double L2Error_Im = x.imag().ComputeL2Error(E_ex_Im, irs,
+                                                  pml->GetMarkedPMLElements());
+
+      ComplexGridFunction x_gf0(fespace);
+      x_gf0 = 0.0;
+      double norm_E_Re = x_gf0.real().ComputeL2Error(E_ex_Re, irs,
+                                                     pml->GetMarkedPMLElements());
+      double norm_E_Im = x_gf0.imag().ComputeL2Error(E_ex_Im, irs,
+                                                     pml->GetMarkedPMLElements());
+
+      cout << " Rel Error - Real Part: || E_h - E || / ||E|| = " << L2Error_Re /
+           norm_E_Re << '\n' << endl;
+      cout << " Rel Error - Imag Part: || E_h - E || / ||E|| = " << L2Error_Im /
+           norm_E_Im << '\n' << endl;
+      cout << " Total Error: " << sqrt(L2Error_Re * L2Error_Re + L2Error_Im *
+                                       L2Error_Im)  << endl;
+   }
+
+   // 16. Save the refined mesh and the solution. This output can be viewed
+   //     later using GLVis: "glvis -m mesh -g sol".
+   {
+      ofstream mesh_ofs("refined.mesh");
+      mesh_ofs.precision(8);
+      mesh->Print(mesh_ofs);
+
+      ofstream sol_r_ofs("sol_r.gf");
+      ofstream sol_i_ofs("sol_i.gf");
+      sol_r_ofs.precision(8);
+      sol_i_ofs.precision(8);
+      x.real().Save(sol_r_ofs);
+      x.imag().Save(sol_i_ofs);
+   }
+
+   // 17. Send the solution by socket to a GLVis server.
+   if (visualization)
+   {
+      string keys;
+      keys = (dim == 3) ? "keys macF\n" : keys = "keys amrRljcUUuu\n";
```

Comment:
Why the random strings?

Reply:
The keys are not really random. They are chosen depending on the problem in order to show the most interesting characteristics of the solution

Reply:
Is this the first example using keys like this in the visualization? If so, it would be informative to explain the usage in comments.